llpc/docs/DdnGraphicsRegisterHandler.md

```diff
@@ -0,0 +1,265 @@
+5. Insert the target `i32` element back to `vec <n x i32>`
+
+## Problems
+
+* Modifying different bits of graphics register (DWORD) multiple times in a naive way may result in inefficient code with redundant instructions, e.g.:
+
+        Call `FUNCTION (1)` for `LOD_BIAS`
+        Call `FUNCTION (1)` for `XY_MAG_FILTER`
+
+    Which results in:
+    1. Extract target element `i32` from `vec <n x i32>`
+    2. Use `amdgcn_ubfe` to extract specified range of bits (Occupied by `LOD_BIAS`)
+    3. A series modification on the symbol of `LOD_BIAS`
+    4. Apply the new symbolic expression of `LOD_BIAS` to the target `i32`
+    5. Insert target `i32` element back to `vec <n x i32>`
+    6. Extract target `i32` element from `vec <n x i32>`
+    7. Use `amdgcn_ubfe` to extract specified range of bits (Occupied by `XY_MAG_FILTER`)
+    8. A series modification to `XY_MAG_FILTER`
+    9. Apply the new symbolic expression of `XY_MAG_FILTER` to the target `i32`
+   10. Insert the target `i32` element back to `vec <n x i32>`
+
+    But this series of manipulation could just be simplify to:
+    1. Extract target element `i32` from `vec <n x i32>`
+    2. Use `amdgcn_ubfe` to extract specified range of bits (Occupied by `LOD_BIAS`)
+    3. A series modification on the symbol of `LOD_BIAS`
+    4. Use `amdgcn_ubfe` to extract specified range of bits (Occupied by `XY_MAG_FILTER`)
+    5. A series modification on the symbol of `XY_MAG_FILTER`
+    6. Merge the new `LOD_BIAS` and `XY_MAG_FILTER` to the target `i32`
+    7. Insert the target `i32` element back to `vec <n x i32>`
+
+* How to avoid calling `amdgcn_ubfe` multiple times for the same range bits.
+
+* If there's actually no modification of `XY_MAG_FILTER` applied to target `i32`, how could we prevent it from inserting the DWORDs back to vec `<n xi32>`.
+
+* For the same register, it may locates differently among GfxIps. Such as `WIDTH`, it locates in `SQ_IMG_RSRC_WORD2` for gfx9, but may locates in `SQ_IMG_RSRC_WORD3` for gfx10. We need to manage these information properly.
+
+# Implementation Overview
+
+We introduce GfxRegHelper for efficiently handling register operations described above.
+
+## Diagram
+### Introduced structs:
+```
+struct BitsInfo
+{
+    uint32_t index;
+    uint32_t offset;
+    uint32_t count;
+};
+```
+```
+struct BitsState
+{
+    Value* pValue = nullptr;
+    bool isModified = false;
+};
+```
+SQ_IMG_SAMP register bits infomation look up table (Gfx9-10)
+```
+static constexpr BitsInfo g_sqImgSampRegBitsGfx9[SqSampRegsCount] =
+{       // Which DWORD      Start bit        Count
+    {            0,            30,             2 }, // FilterMode
+    {            2,            20,             2 }, // XyMagFilter
+    {            2,            22,             2 }, // XyMinFilter
+};
+```
+
+### Phase 1 Bound Registers, BitsState and BitsInfo Look Up Table
+
+* Bound `m_pRegister` to `SQ_IMG_SAMP <4 x i32>`
```

Comment:
`m_pRegister` comes out of nowhere here. Where is it defined? Or is it just a name?

Reply:
IE I don't know if it's important it's a member in some class, or if it's just a variable name.

Reply:
The param is defined in GfxRegHandlerBase,  the name of this parameter might not be that important, but, for preventing reader from doubting where it is defined, i've updated the class name for awareness.
![image](https://user-images.githubusercontent.com/58158709/78059771-08b03400-73bd-11ea-926d-b130a6bf9e05.png)